src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,411 @@
+         * returning a binary encoding after the Base64 decoding is ok when the
+         * user wants PKCS8EncodedKeySpec, generating the key verifies the
+         * binary encoding and allows the KeyFactory to use the provider's
+         * KeySpec()
+         */
+
+        if ((EncodedKeySpec.class).isAssignableFrom(tClass) &&
+            so instanceof Key key) {
+            try {
+                // unchecked suppressed as we know tClass comes from KeySpec
+                // KeyType not relevant here.  We just want KeyFactory
+                if ((PKCS8EncodedKeySpec.class).isAssignableFrom(tClass)) {
+                    so = getKeyFactory(key.getAlgorithm()).
+                        getKeySpec(key, PKCS8EncodedKeySpec.class);
+                } else if ((X509EncodedKeySpec.class).isAssignableFrom(tClass)) {
+                    so = getKeyFactory(key.getAlgorithm())
+                        .getKeySpec(key, X509EncodedKeySpec.class);
+                } else {
+                    throw new IllegalArgumentException("Invalid KeySpec.");
+                }
+            } catch (InvalidKeySpecException e) {
+                throw new IllegalArgumentException("Invalid KeySpec " +
+                    "specified (" + tClass.getName() +") for key (" +
+                    key.getClass().getName() +")", e);
+            }
+        }
+
+        return tClass.cast(so);
+    }
+
+    private KeyFactory getKeyFactory(String algorithm) {
+        try {
+            if (factory == null) {
+                return KeyFactory.getInstance(algorithm);
+            }
+            return KeyFactory.getInstance(algorithm, factory);
+        } catch(GeneralSecurityException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    // Convenience method to avoid provider getInstance checks clutter
+    private CertificateFactory getCertFactory(String algorithm) {
+        try {
+            if (factory == null) {
+                return CertificateFactory.getInstance(algorithm);
+            }
+            return CertificateFactory.getInstance(algorithm, factory);
+        } catch (GeneralSecurityException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Configures and returns a new {@code PEMDecoder} instance from the
+     * current instance that will use KeyFactory and CertificateFactory classes
+     * from the specified {@link Provider}.  Any errors using the
+     * {@code provider} will occur during decoding.
+     *
+     * <p>If {@code provider} is {@code null}, a new instance is returned with
+     * the default provider configuration.
+     *
+     * @param provider the Factory provider.
+     * @return a new PEM decoder instance.
+     */
+    public PEMDecoder withFactory(Provider provider) {
+        return new PEMDecoder(provider, password);
+    }
+
+    /**
+     * Returns a new {@code PEMDecoder} instance from the current instance
```

Comment:
I find the words "a new instance from a current instance" a little unusual.

It might be simpler and clearer to say "Returns a copy of this PEMDecoder that will decrypt encrypted PEM data such as encrypted private keys with the specified password."

(I don't feel it is necessary to mention that it still also decodes unencrypted PEM - that seems implied to me).